Code/EnginePlugins/AiPlugin/Navigation/Implementation/NavMesh.cpp

```diff
@@ -115,20 +115,39 @@ bool ezAiNavMesh::RequestSector(const ezVec2& vCenter, const ezVec2& vHalfExtent
   return res;
 }
 
-// void ezNavMesh::InvalidateSector(SectorID sectorID)
-//{
-//   auto it = m_Sectors.Find(sectorID);
-//   if (!it.IsValid())
-//     return;
-//
-//   auto& sector = it.Value();
-//
-//   if (sector.m_FlagInvalidate == 0 && (sector.m_FlagUsable == 1 || sector.m_FlagUpdateAvailable == 1))
-//   {
-//     sector.m_FlagInvalidate = 1;
-//     m_InvalidatedSectors.PushBack(sectorID);
-//   }
-// }
+void ezAiNavMesh::InvalidateSector(const ezVec2& vCenter, const ezVec2& vHalfExtents)
+{
+  ezVec2I32 coordMin = CalculateSectorCoord(vCenter.x - vHalfExtents.x, vCenter.y - vHalfExtents.y);
+  ezVec2I32 coordMax = CalculateSectorCoord(vCenter.x + vHalfExtents.x, vCenter.y + vHalfExtents.y);
+
+  coordMin.x = ezMath::Clamp<ezInt32>(coordMin.x, 0, m_uiNumSectorsX - 1);
+  coordMax.x = ezMath::Clamp<ezInt32>(coordMax.x, 0, m_uiNumSectorsX - 1);
+  coordMin.y = ezMath::Clamp<ezInt32>(coordMin.y, 0, m_uiNumSectorsY - 1);
+  coordMax.y = ezMath::Clamp<ezInt32>(coordMax.y, 0, m_uiNumSectorsY - 1);
+
+  for (ezInt32 y = coordMin.y; y <= coordMax.y; ++y)
+  {
+    for (ezInt32 x = coordMin.x; x <= coordMax.x; ++x)
+    {
+      InvalidateSector(CalculateSectorID(ezVec2I32(x, y)));
+    }
+  }
+}
+
+void ezAiNavMesh::InvalidateSector(SectorID sectorID)
+{
+  auto it = m_Sectors.Find(sectorID);
+  if (!it.IsValid())
+    return;
+
+  auto& sector = it.Value();
+
+  if (sector.m_FlagInvalidate == 0 && (sector.m_FlagUsable == 1 || sector.m_FlagUpdateAvailable == 1))
+  {
+    sector.m_FlagInvalidate = 1;
+    m_RequestedSectors.PushBack(sectorID);
```

Comment:
This would recreate the sector right away. Another option would be to just unload it, and only recreate it again on demand when another path search comes in.
Or, a third option would be to make this an option to the Invalidate call, whether it should be recreated right away. Then one can use the Invalidate function to also unload sectors, or at least not uncur overhead for recreation, when it isn't needed at the moment.

I think I would prefer the last option, then the Invalidate function can be used for both.

Reply:
Good idea, I'll do as you say. We should also keep in mind that some users may want to always keep all sectors loaded and up to date and there should be an option for that